lib/src/main/scala/spinal/lib/tools/HDElkDiagramGen.scala

```diff
@@ -0,0 +1,552 @@
+              extraEdge.source=fanOutSonName
+              extraEdge.target=moduleName + "." + fanOutSonName
+              extraEdge.isBus=inIsBus & outIsBus
+              var isContained = false
+              for (thisEdge <- edges) {
+                if (thisEdge.source == extraEdge.source && thisEdge.target == extraEdge.target && thisEdge.label == extraEdge.label)
+                  isContained = true
+              }
+              if(!isContained)
+                edges.add(extraEdge)
+              newEdge.target=fanOutSonName
+
+            }
+            else
+              newEdge.target = moduleName + "." + fanOutSonName
+          }
+          else if (fanOut.getComponent() == net.getComponent()) fanOutIsWrong = true
+          else {
+            var judge = false
+            for (thisNode <- topNode.children) {
+              if (thisNode.labelName == fanOut.getComponent().getName()) {
+                judge = true
+                newEdge.target = fanOut.getComponent().getName() + "." + fanOutSonName
+              }
+            }
+            if (!judge)
+              fanOutIsWrong = true
+          }
+        }
+        else if (everyRegisters.contains(fanOut)) {
+          newEdge.target = fanOutSonName
+        }
+        else fanOutIsWrong = true
+
+        /** Setting and adding Edge */
+        if (clkMap.contains(net.getName())) newEdge.highlight = clkMap(net.getName())
+        else if (clkMap.contains(net.clockDomain.toString())) newEdge.highlight = clkMap(net.clockDomain.toString())
+        newEdge.source = sourceName
+        newEdge.isBus = inIsBus & outIsBus
+        if (newEdge.isBus == 1) {
+          if (findParent(net).getClass.getSimpleName == findParent(fanOut).getClass.getSimpleName) newEdge.label = findParent(fanOut).getClass.getSimpleName
+          else newEdge.label = findParent(net).getClass.getSimpleName + " to " + findParent(fanOut).getClass.getSimpleName
+        }
+        else if (inIsBus == 1) {
+          if (!clkMap.contains(net.getName()))
+            newEdge.label = net.getName()
+        }
+        else if (outIsBus == 1) {
+          if (!clkMap.contains(fanOut.getName()))
+            newEdge.label = fanOut.getName()
+        }
+        var isContained = false
+        for (thisEdge <- edges) {
+          if (thisEdge.source == newEdge.source && thisEdge.target == newEdge.target && thisEdge.label == newEdge.label)
+            isContained = true
+        }
+        if (!netIsWrong && !fanOutIsWrong && !isContained && newEdge.source != "" && newEdge.target != "")
+          edges.add(newEdge)
+      }
+    }
+  }
+
+  /** Generating HDElk language for Node */
+  private def drawNodes(thisNode: ElkNode): Unit = {
+    pw.write(s"""{id:"${thisNode.labelName}",\n""")
+    if (thisNode.typeName != "") pw.write(s"""type:"${thisNode.typeName}",\n""")
+    if (thisNode.highlight != 0) pw.write(s"""highlight:${thisNode.highlight},\n""")
+    if (thisNode.inPorts.nonEmpty) {
```

Comment:
How about moving this logic into port specific class? so all related logic can be put into that.
Above code seems identical while it is in or out.